docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,277 @@
+the topic, followed by raft elections for all partitions within the
+topic.
+
+These concepts are not visible to applications.  They are optionally visible
+to system administrators optionally:
+- system monitoring telemetry, where we would surface indicators for how many partitions
+  and topics are currently hibernated.
+- tunable configuration properties, where they could control the delay til hibernation
+  if needed.
+
+
+# Reference-level explanation
+
+## High level design
+
+The topic table, which stores the definition of the topics & partitions
+on the system, is driven by controller log commands.  New commands
+are added for hibernate and wake.  Nodes in a partition's raft group
+notice the hibernation of the topic, and shut down the raft group.
+
+There is a "hibernating" intermediate state where members of the
+partitions' raft groups are responsible for refusing writes, and
+reporting the highest offset back to the controller so that it can
+reliably store the frozen offset state of the topic when storing
+the controller message that finalizes the hibernation.
+
+Once the topic is hibernated, all its partitions will report a
+fictional leader, so that clients know where to send metadata requests,
+and so that monitoring systems do not incorrectly flag the partition
+as being in a problematic state.
+
+On startup, a hibernated partition's storage resources are not
+recovered or replayed.  Hibernated partitions are not subject to
+continuous retention policies, BUT the controller can notice if
+a partition has been hibernated for longer than its retention.ms
+limit, and wake it up for long enough to do housekeeping.
+
+We should still be able to serve metadata requests for idle topics,
+because otherwise any monitoring system will end up re-awakening
+the topics whenever querying them.
+
+## How do we decide when to hibernate?
+
+We don't persist an `atime` type stat for partitions, and rightly so,
+because otherwise we'd be writing metadata every time someone did
+a read.
+
+Detecting an idle partition at runtime therefore relies on a partition
+staying up with stable leadership long enough to reach the idle timeout.
+That's fine if the timeout is a few hours long (if we're changing leaders
+more often than that then we've got bigger problems) but more problematic
+if we have a long idle timeout (perhaps a day or more) with a reasonable
+chance of leadership changes or node restarts before we reach it.
+
+If we wanted to support very long idle timeouts better, we could
+do so by adding something like a kvstore record of the access time
+that is updated at most every N minutes, and when a leader is elected
+it primes its access time from that if it is the same node that was
+also the previous leader.
+
+## What resources are we saving by hibernating?
+
+- Avoid creating new files on log roll during startup
+- Avoid replaying segments during startup
+- No memory overhead from storing segment indices, list of segments, etc.
```

Comment:
If these are backed by a fault-tolerant storage log / structure, can we just demand load what we need on a per-operation basis?

Reply:
The index part yes: loading indices on demand is something we'll probably do anyway to handle partitions with long histories.

For the startup stuff, loading on demand is harder, because we might need to go through replay (of the latest segment) to find the head of the log and be ready to serve metadata requests.  So an on-demand mechanism there would look like a more localized version of the hibernate/awake cycle.